HBR-Proportional-Valve/HBR-Proportional-Valve.ino

```diff
@@ -2,151 +2,132 @@
+struct ValvePin {
+  int pos;
+  int neg;
+};
+
+// Indexes for joystick-related arrays
+const int axisX = 0;
+const int axisY = 1;
+
+// Set up an array of joystick axes.  Yes, a 2-element array.  
+// This helps make the rest of the code more readable
+// and it reduces our need to copy paste code.
+ArduinoJoystick joystick[] = {
+  ArduinoJoystick(0, PIN_JOY_X), 
+  ArduinoJoystick(1, PIN_JOY_Y)
+}; 
+
+// Set up the button which will control start/stop
+ArduinoButton enableButton(2, PIN_EN_BTN);  
+
+// Set up the valves as an array of structs.  
+// Again, a 2-element array. 
+ValvePin valve[] = {
+  { PIN_VALVE_X_POS, PIN_VALVE_X_NEG },
+  { PIN_VALVE_Y_POS, PIN_VALVE_Y_NEG },  
+};
+
+
+// Perform all initial setup for the joystick axes
+void initJoystickAxes() {
+
+  // set up the transform from the input pots to the output PWM signals for the valves
+  LinearTransform* joystickPotTransform = new LinearTransform(-128, 126, -512, 512);
 
-//declare the enable button pin
-const int16_t EN_BTN_PIN = 10;
-
-//declare the output pins (must be PWM capable) for solenoid control
-const int16_t VALVE_X_POS = 3;
-const int16_t VALVE_X_NEG = 5;
-const int16_t VALVE_Y_POS = 6;
-const int16_t VALVE_Y_NEG = 9;
-
-
-//setup the two joystick axis
-ArduinoJoystick joystickX(0, 0);
-ArduinoJoystick joystickY(1, 1);
-
-//setup the button to start/stop
-ArduinoButton enableButton(2, 10);
-
-//setup the transform from the input pots to the output PWM signals for the valves
-LinearTransform transform(-128, 126, -512, 512);
-
-int16_t xValue = 0;
-int16_t yValue = 0;
-
-
-int pwmPins[] = {VALVE_X_POS, VALVE_X_NEG, VALVE_Y_POS, VALVE_Y_NEG};
-
-void setup()
-{
-  Serial.begin(9600);
   
   //calibrate for joystick found here: http://uk.rs-online.com/web/p/products/8430838/
-  //set Low/Center/High positions
-  joystickX.setPoints(390, 520, 640);
-  joystickY.setPoints(390, 520, 640);
-
-  //set the deadbands
-  joystickX.setDeadbands(5, 5, 5);
-  joystickY.setDeadbands(5, 5, 5);
-
-  //set the minimum change to trigger an event
-  joystickX.setThreshold(5);
-  joystickY.setThreshold(5);
-
-  //set the transform from input values to output values
-  joystickX.setTransformation(&transform);
-  joystickY.setTransformation(&transform);
-
-  //get initial readings from the joystick axis
-  xValue = constrain(joystickX.getValue(), -255, 255);
-  yValue = constrain(joystickY.getValue(), -255, 255);
-
-
-  //set all the pins as outputs and initialize to off
-  for (int i = 0; i < 4; i++)
-  {
-    pinMode(pwmPins[i], OUTPUT);
-    analogWrite(pwmPins[i], 0);
+  for (int axis = axisX; axis <= axisY; ++axis) {
+    joystick[axis].setPoints(390, 520, 640); //set Low/Center/High positions
+    joystick[axis].setDeadbands(5, 5, 5);    //set the deadbands
+    joystick[axis].setThreshold(5);          //set the minimum change to trigger an event
+ 
+    //set the transform from input values to output values
+    joystick[axis].setTransformation(joystickPotTransform);
```

Comment:
I would not do this, calibrations should be done per joystick rather than assuming both joysticks to be identical.

I have a pair of these joysticks and they do not share an identical calibration.